### ISSUE TITLE
Passkeys support

### ISSUE THREAD

**Issue by mposolda** [keycloak/keycloak#23656]
### Description

This includes passkeys support in Keycloak including:
- Check it works and implement it
- Automated tests
- Documentation and maybe a blog

### Discussion

e

**Comment by mposolda** [comment]
Preliminary adding to Keycloak 23

**Comment by tnorimat** [comment]
@mposolda I will work on the issue.
I think it might be better to create an issue for the first passkey support as PREVIEW.
What do you think about that? If OK, I will create the issue.

**Comment by mposolda** [comment]
Postponing to Keycloak 24 for now. We have some work related to passkeys already done in Keycloak 23, but some more things (especially including automated tests) we likely may not be able to cover, so postponing this.

**Comment by Utopiah** [comment]
> Passkeys are a super hot topic these days. 

Indeed, I discovered Keycloak by setting up a WebAuthN flow (for WebXR) and listening just this morning to https://www.theverge.com/24071753/passkeys-password-manager-wearables-vergecast (as a habit) and my very first thought was "Wait a minute... can I use Keycloak for this?!".

I found this issue via https://github.com/keycloak/keycloak/discussions/16201 , quite eager to see it landing on 24.

PS: taking the occasion to thank you for the great work!

**Comment by tnorimat** [comment]
@mposolda I replaced "Resident Key" (defined by WebAuthn Level 2) with "Discoverable Credential" (defined by WebAuthn Level 3) by https://github.com/keycloak/keycloak/pull/27104 .
How about adding it to this epic issue? Also could you check the PR?

**Comment by jeetparte** [comment]
FYI I found two addressed issues related to Passkeys which are not tracked here — https://github.com/keycloak/keycloak/issues/27147 and https://github.com/keycloak/keycloak/issues/27105.

Can we link them here as well?

Although the above items are mentioned in the Release Notes, it's easier to just visit this page to keep track of all the progress being made on Passkeys implementation (plus unlike release notes, updates here are visible to us users between releases).

BTW thank you for all the work being done to support this Passkeys feature. 🙏

**Comment by mposolda** [comment]
@jeetparte Thanks, I've linked both issues. Also linked some bugs found (After first version of Passkeys Conditional UI Authenticator was done): https://github.com/keycloak/keycloak/issues/29585 and https://github.com/keycloak/keycloak/issues/29586 and also the issue https://github.com/keycloak/keycloak/issues/25779

**Comment by rmartinc** [comment]
Just to register this note somewhere. More or less what keycloak is doing to include passkeys in the authentication flow is using the idea explained here [Implementing a combination autofill / conditional UI + modal flow](https://developers.yubico.com/WebAuthn/Concepts/Passkey_Autofill/Implementation_Guidance/Modal_and_Autofill_Flow.html). Both conditional (username input) and modal (different button) UIs are added to the login forms. Passkeys will still be preview in 26.3, but more or less all the pieces are now in place. The initial goal is polishing the implementation with the issues found in 26.3 to support the feature in the next 26.4.🤞

**Comment by logopk** [comment]
Unfortunately I cannot get local Passkeys to work with Bitwarden. (KeyCloak 26.3.3)

Safari will not show me an option to use Bitwarden to create a passkey on my MacBook - only using a remote device like my iPhone. This works but is a bit clumsy as I always have to scan the QRCode before I can confirm the user and login.

<img width="475" height="291" alt="Image" src="https://github.com/user-attachments/assets/9864c3a6-d2d9-400e-b034-db7c3b739c09" />

This flow is seamlessly working with passkeys.io.

Anything I can do to make this work? ChatGPT tried to lead me to config options that are not (no longer) available in the keycloak admin console...

> 1. Safari is picky about platform credentials
> Safari 17+ (including 18) generally favors synced iCloud passkeys. It may not offer local-only platform credentials unless:
> The relying party (Keycloak) sets the right WebAuthn flags.
> The authenticatorAttachment is platform, not cross-platform.
> The residentKey is set to required or preferred.
> Apple devices prefer syncing credentials unless you're offline or explicitly using local fallback.
> 2. Keycloak 26 uses Passkey/WebAuthn via FIDO2
> But Keycloak’s default WebAuthn flow may not request resident credentials or may not be clear enough to browsers (especially Safari) that local is desired.
> 
> 
> 🧪 How to Fix / Work Around It
> ✅ 1. Adjust Keycloak WebAuthn settings
> Go to the Keycloak Admin Console → Authentication → WebAuthn Policy. Then:
> Set "User Verification Requirement" to: required
> Set "Authenticator Attachment" to: platform
> Set "Resident Key Requirement" to: required
> Optionally enable: "Require Discoverable Credentials"
> This signals to Safari that you want a local, device-bound passkey, not one that must be synced via iCloud or Bitwarden.
> ✅ 2. Use Chrome or Firefox to register passkey
> During initial registration, use a browser that gives you full control (Chrome/Edge/Firefox), then test login with Safari.
> Bitwarden and Safari both store passkeys differently:
> 
> Bitwarden stores synced passkeys (cross-platform).
> Safari prefers iCloud-linked ones. If local isn't available, it defaults to iCloud.
> You can test with macOS’s WebAuthn debug console:
> sudo defaults write com.apple.WebAuthn debugMode -bool true
> Then open Safari → DevTools → WebAuthn tab.

"Resident Key Requirement" is not available in the UI.

Any thoughts on that? Chrome is doing the same btw ....

**Comment by rmartinc** [comment]
@logopk The webauthn policy options in the console are in `Authentication -> Policies -> Webauthn Passwordless Policy` (this is the policy used by passkeys). In that policy you have the settings you are commenting about.

**Comment by logopk** [comment]
There is no "Resident Key Requirement" only "Require discoverable credential".

**Comment by rmartinc** [comment]
"Resident Key Requirement" set to "required" is the same than "Require discoverable credential" to Yes (that should be configured by default if you are using the last version).

**Comment by logopk** [comment]
@rmartinc ok that's good to know - however the popup of bitwarden does not come up only the screenshot above.

<img width="473" height="272" alt="Image" src="https://github.com/user-attachments/assets/7095a191-b26a-450b-b138-f52203705126" />

There isn't even an option for the remote passkeys anymore.

**Comment by mposolda** [comment]
Closed as all the required issues are done. Passkeys feature support will be available in Keycloak 26.4.0.

Anyone welcome to try passkeys in latest [Keycloak nightly build ](https://www.keycloak.org/nightly/) and provide feedback and eventually create additional issues if some bugs found or if you have ideas for improvements. See also [latest nightly docs](https://www.keycloak.org/docs/nightly/server_admin/index.html)
